src/serac/physics/integrators/hyperelastic_traction_integrator.cpp

```diff
@@ -0,0 +1,201 @@
+// Copyright (c) 2019-2021, Lawrence Livermore National Security, LLC and
+// other Serac Project Developers. See the top-level LICENSE file for
+// details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+
+#include "serac/physics/integrators/hyperelastic_traction_integrator.hpp"
+
+namespace serac::mfem_ext {
+
+void HyperelasticTractionIntegrator::AssembleFaceVector(const mfem::FiniteElement& el1, const mfem::FiniteElement&,
+                                                        mfem::FaceElementTransformations& Tr, const mfem::Vector& elfun,
+                                                        mfem::Vector& elvec)
+{
+  int dim = el1.GetDim();
+  int dof = el1.GetDof();
+
+  shape_.SetSize(dof);
+  elvec.SetSize(dim * dof);
+
+  DSh_u_.SetSize(dof, dim);
+  DS_u_.SetSize(dof, dim);
+  J0i_.SetSize(dim);
+  F_.SetSize(dim);
+  Finv_.SetSize(dim);
+
+  PMatI_u_.UseExternalData(elfun.GetData(), dof, dim);
+
+  int intorder = 2 * el1.GetOrder() + 3;
+
+  const mfem::IntegrationRule& ir = mfem::IntRules.Get(Tr.FaceGeom, intorder);
+
+  elvec = 0.0;
+
+  mfem::Vector trac(dim);
+  mfem::Vector ftrac(dim);
+  mfem::Vector nor(dim);
+  mfem::Vector fnor(dim);
+  mfem::Vector u(dim);
+  mfem::Vector fu(dim);
+
+  for (int i = 0; i < ir.GetNPoints(); i++) {
+    const mfem::IntegrationPoint& ip = ir.IntPoint(i);
+    mfem::IntegrationPoint        eip;
+    Tr.Loc1.Transform(ip, eip);
+
+    Tr.Face->SetIntPoint(&ip);
+
+    // Compute traction
+    function_.Eval(trac, *Tr.Face, ip);
+
+    if (!compute_on_reference_) {
+      Tr.Elem1->SetIntPoint(&eip);
+      CalcInverse(Tr.Elem1->Jacobian(), J0i_);
+
+      el1.CalcDShape(eip, DSh_u_);
+      Mult(DSh_u_, J0i_, DS_u_);
+      MultAtB(PMatI_u_, DS_u_, F_);
+
+      for (int d = 0; d < dim; d++) {
+        F_(d, d) += 1.0;
+      }
```

Comment:
might be less of an issue for finite deformation, but there can a considerable loss of precision by actually adding 1.0 to the diagonal here. We should be able to calculate any strain measures from `(F - I)` directly, avoiding any loss of precision.

Reply:
Actually, it turns out it is best to calculate du/dx and add I to it than to calculate dx/DX

When u is small and x is large, then x~X so dx/DX looks like I, but you could still have du/dX large

Reply:
Yeah, I think I kept the material interface with the deformation gradient instead of the displacement gradient (F-I) to be consistent with the existing MFEM material models. @solberg2 , what do the material models in your codes use?

Reply:
So what we do is Calculate du/dX and then calculate F = du/dX + I.  That's more accurate than simply calculating dx/DX
Also it means you still have du/dX lying around if you want it.  For hypoelastic materials you want du/dx of course, but with x calculated at the midpoint configuration.

Reply:
I'm not advocating for calculating dx/DX, just observing that you rarely ever _need_ F. When u is small, actually forming F will always result in a loss of precision. If a constitutive model needs a strain measure, we can just calculate those directly from du_dx without losing precision by forming F as an intermediate (even though most strain measures are typically expressed most succinctly in terms of F, with exact arithmetic).

Reply:
Yeah.  That's why we essentially form both and pass them down to the material model.  Hypoelastic models based on the Green-Naghdi formulation need F_{n+1} and F_{n} because they calculate R_{n+1} and R_{n} so they can rotate back and forth to the reference configuration

Reply:
So what Diablo does is create different "kinematics" modules

   Hyperelastic
   Hypoelastic
   
 That's kinda of a low-level approach of what we do with "fields", where we let each element level request which fields it wants
 
 To some extent I think what would make sense is to have a kinematics module that has the routines to calculate a bunch of different kinematic variables  (F_{n+1}, du/dx, R_{n+1}, )

Reply:
I think this approach will be useful for @samuelpmishLLNL 's upcoming rewrite.